lightning/src/chain/mod.rs

```diff
@@ -57,6 +57,81 @@ pub trait Listen {
+/// notification, it would pass along the relevant transactions using this interface.
+///
+/// The intended use is as follows:
+/// - Call [`transactions_confirmed`] for any on-chain activity of interest.
+/// - Return such transactions from [`get_relevant_txids`] until they have enough confirmations to
+///   no longer be in danger of chain reorganization.
+/// - Call [`transaction_unconfirmed`] for any transaction returned by [`get_relevant_txids`] that
+///   has been reorganized out of the chain.
+/// - Call [`best_block_updated`] whenever a new chain tip becomes available.
+///
+/// See individual method documentation for further details.
+///
+/// [`transactions_confirmed`]: Self::transactions_confirmed
+/// [`transaction_unconfirmed`]: Self::transaction_unconfirmed
+/// [`best_block_updated`]: Self::best_block_updated
+/// [`get_relevant_txids`]: Self::get_relevant_txids
+pub trait Confirm {
+	/// Processes transactions confirmed in a block with a given header and height.
+	///
+	/// Should be called for any transactions registered by [`Filter::register_tx`] or any
+	/// transactions spending an output registered by [`Filter::register_output`]. Such transactions
+	/// appearing in the same block do not need to be included in the same call; instead, multiple
+	/// calls with additional transactions may be made.
+	///
+	/// May be called before or after [`best_block_updated`] for the corresponding block. However,
+	/// it must not be called after [`best_block_updated`] if `header` corresponds to a block that
+	/// has been reorganized from the chain (i.e., if the transaction is no longer confirmed).
+	///
+	/// [`best_block_updated`]: Self::best_block_updated
+	fn transactions_confirmed(&self, header: &BlockHeader, txdata: &TransactionData, height: u32);
+
+	/// Processes a transaction that is no longer confirmed as result of a chain reorganization.
+	///
+	/// Should be called for any transaction returned by [`get_relevant_txids`] if it has been
+	/// reorganized out of the best chain. Once called, the given transaction should not be returned
+	/// by [`get_relevant_txids`] unless it has been reconfirmed via [`transactions_confirmed`].
+	///
+	/// [`get_relevant_txids`]: Self::get_relevant_txids
+	/// [`transactions_confirmed`]: Self::transactions_confirmed
+	fn transaction_unconfirmed(&self, txid: &Txid);
+
+	/// Processes an update to the best header connected at the given height.
+	///
+	/// Should be called when a new header is available but may be skipped for intermediary blocks
+	/// if they become available at the same time.
+	///
+	/// Implementations are not required to handle chain reorganizations unless `height` is not
+	/// greater than the height given in the previous call. Instead, reorganizations should be
+	/// handled by [`transaction_unconfirmed`].
+	///
+	/// [`transaction_unconfirmed`]: Self::transaction_unconfirmed
+	fn best_block_updated(&self, header: &BlockHeader, height: u32);
```

Comment:
Is a user require to call `best_block_updated` if any transaction has been `transaction_confirmed` for the associated height and hash ? I believe If it's a MUST just say it.

Reply:
If it's not necessary can you say it explicitly in either `transaction_confirmed` or `best_block_updated` ? It wasn't obvious to me :)

Reply:
Added here.

Reply:
Hmm, its easy to confuse the new wording for "if you passed a header in via `transactions_confirmed`, don't bother passing it in here, too" which I don't think is what we want.  Instead we just mean "if you pass a header in via `transactions_confirmed` but by the time you go to call `best_block_updated` the best block is different, just call this with the latest one", which is pretty different.

Reply:
Alright, I'm just going to drop it to avoid confusion. Let me know if the docs around use of `best_block_updated` can be clarified in any other way. It's currently mentioned in `transactions_confirmed` and in the trait-level docs.